packages/models-tree/src/models-tree/ModelsTreeDefinition.ts

```diff
@@ -740,40 +774,72 @@ async function createInstanceKeyPathsFromTargetItems({
-      from(ids.subjects).pipe(mergeMap((id) => createSubjectInstanceKeysPath(id, idsCache))),
-      from(ids.models).pipe(mergeMap((id) => createModelInstanceKeyPaths(id, idsCache))),
-      from(ids.categories).pipe(mergeMap((id) => createCategoryInstanceKeyPaths(id, idsCache))),
-      from(elementBlocks).pipe(mergeMap((block) => createGeometricElementInstanceKeyPaths(imodelAccess, idsCache, hierarchyConfig, block))),
+  return lastValueFrom(
+    from(targetItems).pipe(
+      releaseMainThreadOnItemsCount(2000),
+      mergeMap(async (key): Promise<{ key: string; type: number } | { key: ElementsGroupInfo; type: 0 }> => {
+        if ("parent" in key) {
+          return { key, type: 0 };
+        }
+
+        const subjectCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.Subject");
+        const isSubject = subjectCheck instanceof Promise ? await subjectCheck : subjectCheck;
+        if (isSubject) {
+          return { key: key.id, type: 1 };
+        }
+
+        const modelCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.Model");
+        const isModel = modelCheck instanceof Promise ? await modelCheck : modelCheck;
+        if (isModel) {
+          return { key: key.id, type: 2 };
+        }
+
+        const spatialCategoryCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.SpatialCategory");
+        const isSpatialCategoryCheck = spatialCategoryCheck instanceof Promise ? await spatialCategoryCheck : spatialCategoryCheck;
+        if (isSpatialCategoryCheck) {
+          return { key: key.id, type: 3 };
+        }
+
+        return { key: key.id, type: 0 };
+      }, 2),
+      toArray(),
+      switchMap(async (keysParsed) => {
+        const ids = {
+          models: new Array<Id64String>(),
+          categories: new Array<Id64String>(),
+          subjects: new Array<Id64String>(),
+          elements: new Array<Id64String | ElementsGroupInfo>(),
+        };
+        for (const keyParsed of keysParsed) {
+          if (keyParsed.type === 1) {
+            ids.subjects.push(keyParsed.key);
+            continue;
+          }
+          if (keyParsed.type === 2) {
+            ids.models.push(keyParsed.key);
+            continue;
+          }
+          if (keyParsed.type === 3) {
+            ids.categories.push(keyParsed.key);
+            continue;
+          }
+          ids.elements.push(keyParsed.key);
+        }
+
+        const elementsLength = ids.elements.length;
+        return collect(
+          merge(
+            from(ids.subjects).pipe(mergeMap((id) => createSubjectInstanceKeysPath(id, idsCache))),
+            from(ids.models).pipe(mergeMap((id) => createModelInstanceKeyPaths(id, idsCache))),
+            from(ids.categories).pipe(mergeMap((id) => createCategoryInstanceKeyPaths(id, idsCache))),
+            from(ids.elements).pipe(
+              bufferCount(Math.ceil(elementsLength / Math.ceil(elementsLength / 5000))),
+              releaseMainThreadOnItemsCount(2),
```

Comment:
Does this `releaseMainThreadOnItemsCount(2)` give any value? As far as I can tell, `createGeometricElementInstanceKeyPaths` doesn't do anything expensive, so it shouldn't be a problem if we call it many times from the same microtask. 

What could be a problem though, is `createGeometricElementInstanceKeyPaths` sending out `queryRows` requests - we don't want to send out too many of them at the same time. Maybe `mergeMap` should have a concurrency limit of ~10 to avoid that?

Reply:
You were right, removed